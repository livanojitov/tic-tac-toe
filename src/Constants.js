const EMPTY = 0;
const COMPUTER = 1;
const USER = 2;
const TOTAL_SQUARES = 9;
const RANDOM = 1;
const HARD = 2;
const LOST = 0;
const WON = 1;
const DRAW = 2;
const UNKNOWN = 3;
const ENGLISH = 0;
const FRENCH = 1;
const SPANISH = 2;
const IMAGES_PER_CATEGORY = 20;

export {
  EMPTY,
  COMPUTER,
  USER,
  RANDOM,
  HARD,
  TOTAL_SQUARES,
  LOST,
  WON,
  DRAW,
  UNKNOWN,
  ENGLISH,
  FRENCH,
  SPANISH,
  IMAGES_PER_CATEGORY,
};
