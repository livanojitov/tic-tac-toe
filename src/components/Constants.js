const EMPTY = 0;

const COMPUTER = 1;
const USER = 2;

const EASY = 1;
const HARD = 2;
const NORMAL = 3;

const TOTAL_SQUARES = 9;

const CATEGORY = 'Category';

const YOU = 'You';
const OPPONENT = 'Computer';
const REFRESH = 'Refresh';
const USER_IMAGE_TOOLTIP = 'You will play with this image';
const COMPUTER_IMAGE_TOOLTIP = 'The computer will play with this image';
const REFRESH_TOOLTIP = 'Click to refresh the two images';

const START = 'Start';
const YES = 'Yes';
const NO = 'No';
const START_TOOLTIP = 'Who starts playing the game: You or the computer?';
const YES_TOOLTIP = 'You start playing the game.';
const NO_TOOLTIP = 'The computer starts playing the game.';
const GAME_STARTED_BY = 'Game started by:';

const LEVEL = 'Level';
const COMPUTER_PLAYS_EASY = 'Easy';
const COMPUTER_PLAYS_HARD = 'Hard';
const COMPUTER_PLAYS_NORMAL = 'Normal';
const LEVEL_TOOLTIP = 'How much thinking the computer does before playing.';
const COMPUTER_PLAYS_EASY_TOOLTIP = 'The computer plays randomly.';
const COMPUTER_PLAYS_HARD_TOOLTIP = 'The computer does some thinking before playing.';
const COMPUTER_PLAYS_NORMAL_TOOLTIP = 'The computer sometimes plays easy and sometimes plays hard.';
const LEVEL_HISTORY_EASY = 'Level: Easy';
const LEVEL_HISTORY_HARD = 'Level: Hard';
const LEVEL_HISTORY_NORMAL = 'Level: Normal';

const PLAY = 'Start the game';
const GAME_OVER = 'Game over';
const PLAY_AGAIN = 'Play again';
const LOST = 'You lost!';
const WON = 'You won!';
const DRAW = 'It\'s a draw!';

const GAME_WRITTEN_WITH = 'This game is written with:';
const SOURCE_CODE = 'Source Code:';
const OTHER_GAMES = 'Other games:';

const AUTHOR = 'Author:';
const EMAIL = 'Email:';
const GITHUB = 'Github:';
const LINKEDIN = 'LinkedIn:';
const HIRE_ME = 'Hire me!';

const NO_HISTORY_YET = 'No history yet. Play some games and comeback.';

const HOME = 'Home';
const HISTORY = 'History';
const ABOUT = 'About';
const CONTACT = 'Contact';

export { EMPTY, COMPUTER, USER, EASY, HARD, NORMAL, TOTAL_SQUARES, CATEGORY, YOU, OPPONENT, REFRESH, USER_IMAGE_TOOLTIP, COMPUTER_IMAGE_TOOLTIP, REFRESH_TOOLTIP, START, YES, NO, START_TOOLTIP, YES_TOOLTIP, NO_TOOLTIP, GAME_STARTED_BY, LEVEL, COMPUTER_PLAYS_EASY, COMPUTER_PLAYS_HARD, LEVEL_TOOLTIP, COMPUTER_PLAYS_EASY_TOOLTIP, COMPUTER_PLAYS_HARD_TOOLTIP, LEVEL_HISTORY_EASY, LEVEL_HISTORY_HARD, PLAY, GAME_OVER, PLAY_AGAIN, LOST, WON, DRAW, GAME_WRITTEN_WITH, SOURCE_CODE, OTHER_GAMES, AUTHOR, EMAIL, GITHUB, LINKEDIN, HIRE_ME, NO_HISTORY_YET, HOME, HISTORY, ABOUT, CONTACT, COMPUTER_PLAYS_NORMAL, COMPUTER_PLAYS_NORMAL_TOOLTIP, LEVEL_HISTORY_NORMAL };  
