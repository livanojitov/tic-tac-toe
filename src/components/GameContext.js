import React, { useState, createContext, useEffect } from 'react';
export const GameContext = createContext();

const GameContextProvider = (props) => {

  const [games, setGames] = useState([]);

  useEffect(() => {
    if (typeof(Storage) !== "undefined") {
      if (games.length){
        localStorage.tictactoe = JSON.stringify(games);
      }else{
        if (GameContextProvider.delete === 1){
          GameContextProvider.delete = 0;
          localStorage.removeItem('tictactoe');
        }else{
          if (localStorage.tictactoe){
            setGames(JSON.parse(localStorage.tictactoe));
          }
        }
      }
    }
  },[games]);

  const addGame = (game) => {
    game.id = Math.floor(Math.random() * 10000);
    setGames([...games, game]);
  };

  const removeGame = (id) => {
    if (games.length > 0){
      GameContextProvider.delete = 1;
      id = id * 1;
      setGames(games.filter(game => game.id !== id));   
    } 
  }

  const getGame = (id) => {
    if (games.length > 0){
      id = id * 1;
      return games.filter((game) => game.id === id)[0];
    }else{
      return false;
    }    
  }

  return (
    <GameContext.Provider value={{ games, addGame, removeGame, getGame }}>
      {props.children}
    </GameContext.Provider>
  );

}
 
export default GameContextProvider;