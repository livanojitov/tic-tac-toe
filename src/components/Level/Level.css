.level{
  display: inline-block;
  width: 45%;  
}

.level select{
  margin-left: 2%;
}

.level span{
  cursor: default;
}