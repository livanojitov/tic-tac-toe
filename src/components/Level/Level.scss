.level {
  display: inline-block;
  width: 45%;

  select {
    margin-left: 2%;
  }

  span {
    cursor: default;
  }
}
