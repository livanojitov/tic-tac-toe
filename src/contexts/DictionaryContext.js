import React, { createContext } from "react";
import useFetchData from "../hooks/useFetchData";
import useLanguage from "../hooks/useLanguage";
import * as constants from "../Constants";

export const DictionaryContext = createContext();

const DictionaryContextProvider = ({ children }) => {
  let { data: dictionary, error } = useFetchData("dictionary");

  const [language, changeLanguageHandler] = useLanguage(constants.ENGLISH);

  if (error) {
    return (
      <div className="error" data-testid="error">
        {error}
      </div>
    );
  }

  dictionary = dictionary[language];

  return (
    <DictionaryContext.Provider value={{ dictionary, changeLanguageHandler }}>
      {children}
    </DictionaryContext.Provider>
  );
};

export default DictionaryContextProvider;
