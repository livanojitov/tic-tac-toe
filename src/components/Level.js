import React,{ Component } from 'react';
import { HistoryContext }   from '../contexts/HistoryContext';
import * as constants from './Constants';

const { LEVEL, LEVEL_TOOLTIP, EASY, HARD, NORMAL, COMPUTER_PLAYS_EASY, COMPUTER_PLAYS_HARD, COMPUTER_PLAYS_NORMAL, LEVEL_HISTORY_EASY, LEVEL_HISTORY_HARD, LEVEL_HISTORY_NORMAL} = constants;

class Level extends Component {
  static contextType = HistoryContext;
  state = {
    level : EASY
  }

  render() {
    const { history } = this.context;
    const { disabled, level} = this.props;

    const html = (!history) ? ( 
           
      <div className="level">

        <span title = {LEVEL_TOOLTIP}>{LEVEL}?</span>

        <select onChange={this.setLevel} disabled = {disabled} >
          <option key={EASY}   value={EASY}>{COMPUTER_PLAYS_EASY}</option>  
          <option key={HARD}   value={HARD}>{COMPUTER_PLAYS_HARD}</option>  
          <option key={NORMAL} value={NORMAL}>{COMPUTER_PLAYS_NORMAL}</option>  
        </select>

      </div>
    ) : (
      <div className="level">
        {level === EASY ? LEVEL_HISTORY_EASY : ((level === HARD) ? LEVEL_HISTORY_HARD : LEVEL_HISTORY_NORMAL)}
      </div>
    );

    return (
      html
    )  
  }

  setLevel = (e) => {
    const level = e.target.value * 1;
    this.setState(() => ({level}));
  }

  onLevelChange = () => {
    if (this.props.onLevelChange){
      this.props.onLevelChange(this.state.level);
    }
  }

  componentDidUpdate(prevProps, prevState){
    if (prevState.level !== this.state.level){
      this.onLevelChange();
    }     
  }

  componentDidMount(){
    this.onLevelChange();
  }  
} 

export default Level;