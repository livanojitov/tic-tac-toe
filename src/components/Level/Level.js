import React, { useContext, useMemo } from "react";
import { DictionaryContext } from "../../contexts/DictionaryContext";
import { GameContext } from "../../contexts/GameContext";
import { changeLevel } from "../../store/game/gameActions";
import { RANDOM, HARD } from "../../Constants";
import "./Level.scss";

const Level = () => {
  const { dictionary } = useContext(DictionaryContext);
  const { game, dispatch } = useContext(GameContext);

  return useMemo(() => {
    return (
      <div className="level" data-testid="level">
        <span title={dictionary.LEVEL_TOOLTIP}>{dictionary.LEVEL}?</span>
        <select
          data-testid="select"
          defaultValue={game.level}
          onChange={(e) => {
            dispatch(changeLevel(e.target.value * 1));
          }}
          disabled={!game.disabled}
        >
          <option data-testid="select-option" key={RANDOM} value={RANDOM}>
            {" "}
            {dictionary.LEVEL_EASY}
          </option>
          <option data-testid="select-option" key={HARD} value={HARD}>
            {" "}
            {dictionary.LEVEL_HARD}
          </option>
        </select>
      </div>
    );
  }, [dictionary, game.disabled, game.level, dispatch]);
};

export default Level;
