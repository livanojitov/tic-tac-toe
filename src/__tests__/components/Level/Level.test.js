import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom/extend-expect";
import { DictionaryContext } from "../../../contexts/DictionaryContext";
import { GameContext } from "../../../contexts/GameContext";
import Level from "../../../components/Level/Level";
import { RANDOM } from "../../../Constants";

describe("Level", () => {
  let dictionary;
  let game;
  let dispatch;

  beforeEach(() => {
    dictionary = {
      LEVEL: "Level",
      LEVEL_EASY: "Easy",
      LEVEL_HARD: "Hard",
      LEVEL_NORMAL: "Normal",
      LEVEL_TOOLTIP: "How much thinking the computer does before playing.",
    };

    game = {
      level: RANDOM,
      disabled: true,
    };

    dispatch = jest.fn();
  });

  const renderLevel = () => {
    return render(
      <DictionaryContext.Provider value={{ dictionary }}>
        <GameContext.Provider value={{ game, dispatch }}>
          <Level />
        </GameContext.Provider>
      </DictionaryContext.Provider>
    );
  };

  it("renders correctly", () => {
    const { asFragment } = renderLevel();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display the level dropdown", () => {
    renderLevel();

    expect(screen.getByTestId("level")).not.toBeNull();
    expect(screen.getByTestId("level")).toHaveClass("level");
  });

  it("should change the computer playing level in the dropdown", () => {
    renderLevel();

    expect(
      screen.getByRole("option", { name: dictionary.LEVEL_EASY }).selected
    ).toBeTruthy();
    expect(
      screen.getByRole("option", { name: dictionary.LEVEL_HARD }).selected
    ).toBeFalsy();

    userEvent.selectOptions(
      screen.getByTestId("select"),
      screen.getByText(dictionary.LEVEL_HARD)
    );

    expect(
      screen.getByRole("option", { name: dictionary.LEVEL_EASY }).selected
    ).toBeFalsy();
    expect(
      screen.getByRole("option", { name: dictionary.LEVEL_HARD }).selected
    ).toBeTruthy();
  });
});
