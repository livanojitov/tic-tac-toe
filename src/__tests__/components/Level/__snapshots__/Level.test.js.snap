// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Level renders correctly 1`] = `
<DocumentFragment>
  <div
    class="level"
    data-testid="level"
  >
    <span
      title="How much thinking the computer does before playing."
    >
      Level?
    </span>
    <select
      data-testid="select"
    >
      <option
        data-testid="select-option"
        selected=""
        value="1"
      >
         Easy
      </option>
      <option
        data-testid="select-option"
        value="2"
      >
         Hard
      </option>
    </select>
  </div>
</DocumentFragment>
`;
